### PR TITLE
Fill api token form when user arrives from manage project page

### PR DESCRIPTION
**Pull request by ristomcgehee**
This PR will make it so that when the user is on the manage project page, and they click "Create a token for <project_name>", it will have that project selected on the "Add API token" page.

I used query parameters to pass the project name, and I noticed the query parameters were not getting propagated through the reauth form, so I modified that to pass the query parameters.

Closes https://github.com/pypi/warehouse/issues/6606

### CODE REVIEW

**Comment by di** [comment]
Thanks for working on this @ristomcgehee!

I think one slightly simpler way to go about this would be to make the "Create a token for ..." button submit a POST request to `/manage/account/token/` with the scope already set (but the token name missing).

That would result in the user landing on a page like this:

image

And we wouldn't need to add support for the query parameter, change the reauthentication form, etc.

**Comment by ristomcgehee** [comment]
If we were to change the "Create a token for ..." button from a GET to a POST action, then the user would no longer be able to open that button in a new tab. This would slightly reduce the usability, I think. Also, I don't think it's very user friendly to show an error message to the user right when the page loads. With that said, do you still want the button to be a POST instead of GET?

By the way, I heard the episode you were in on the Sustain podcast. It was cool learning some of the history of PyPI!

**Comment by di** [comment]
> If we were to change the "Create a token for ..." button from a GET to a POST action, then the user would no longer be able to open that button in a new tab. This would slightly reduce the usability, I think. Also, I don't think it's very user friendly to show an error message to the user right when the page loads. With that said, do you still want the button to be a POST instead of GET?

That's a fair point! Let me give this a closer review in that case.

> By the way, I heard the episode you were in on the Sustain podcast. It was cool learning some of the history of PyPI!

Thanks! If you want more history, this talk covers a lot of it: https://www.youtube.com/watch?v=AQsZsgJ30AE

**Comment by ristomcgehee** [comment]
@di just wanted to give a friendly reminder to review this when you have the time.

**Comment by ristomcgehee** [comment]
@miketheman or @ewdurbin, would one of you be able to review this PR?

**Review by miketheman — COMMENTED** [review]
I was poking at the behavior of this code, and noted a couple of things:

1. The project dropdown gets populated correctly from the `?q=` param, cool!
1. The resulting URL after the user has POSTed create token remains with the `?q=` param - which is no longer correct - since now the user could have chosen a different project in the dropdown, yet their URL is still pointing to the previous project name

I didn't get into the reauth behaviors yet, but wanted to provide initial notes.

**Comment by ristomcgehee** [comment]
Thanks for reviewing, Mike.

> The resulting URL after the user has POSTed create token remains with the `?q=` param - which is no longer correct - since now the user could have chosen a different project in the dropdown, yet their URL is still pointing to the previous project name

That's a good catch. I don't see an easy way of removing the `?q=` param from the URL after the user submits the POST create token request, unless we want to change [create_macaroon()](https://github.com/pypi/warehouse/blob/616130795e4fc5ed4bded04af9af8707ef6631a8/warehouse/manage/views.py#L1008) to redirect to a GET at the end of the method. I could change the query name from `q` to `initial_project_selection` to clarify its use. That way I don't think a user would be confused if they chose a different project and then see the previous project name in the URL. Would you like me to make that change?

**Comment by miketheman** [comment]
> unless we want to change [create_macaroon()](https://github.com/pypi/warehouse/blob/616130795e4fc5ed4bded04af9af8707ef6631a8/warehouse/manage/views.py#L1008) to redirect to a GET at the end of the method.

Definitely a good callout - looking at that one, we can see that this was explicitly not converted in #12022 - so I'll consder that out of scope for this changeset.

@ristomcgehee Please resolve the merge conflicts - likely by merging upstream and re-running `make translations` to get the updated line references.

@di tagging you for review

**Review by miketheman — APPROVED** [review]

**Comment by ristomcgehee** [comment]
@di I know this has been open for a while, but I'm still willing to work on this if you review it.

File: warehouse/accounts/views.py [hunk 1]
```diff
@@ -1351,7 +1352,9 @@ def reauthenticate(request, _form_class=ReAuthenticateForm):
 
     if form.next_route.data and form.next_route_matchdict.data:
         redirect_to = request.route_path(
-            form.next_route.data, **json.loads(form.next_route_matchdict.data)
+            form.next_route.data,
+            **json.loads(form.next_route_matchdict.data),
+            _query=json.loads(form.next_route_query.data),
```
Comment (di) [inline warehouse/accounts/views.py:1357]: I think we want to protect against `form.next_route_matchdict.data` potentially containing a `_query` key here. Something like:

```suggestion
            **json.loads(form.next_route_matchdict.data) | dict(_query=json.loads(form.next_route_query.data)),
```

**Comment by ristomcgehee** [comment]
@di Thanks for the review. I've accepted your suggestions, so I'm ready for another review.

**Review by di — APPROVED** [review]

**Comment by di** [comment]
Thanks for your patience here @ristomcgehee!

**Pull request merged**